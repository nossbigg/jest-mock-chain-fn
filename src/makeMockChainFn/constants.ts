export const JestMockChainFnInternalState = Symbol(
  'JestMockChainFn:InternalState'
)
