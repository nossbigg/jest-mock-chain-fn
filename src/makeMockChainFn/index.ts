export * from './makeMockChainFn'
export * from './typedefs'
export * from './constants'
