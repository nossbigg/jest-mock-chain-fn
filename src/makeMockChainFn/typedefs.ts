export type BulilderMockInternalStateType = {
  mocks: Record<string, jest.Mock>
  calls: CallsType[]
}

export type MakeBuilderMockOptions = {
  mockPropertyReturns?: Record<string, MockPropertyReturnsEntry>
}
export type MockPropertyReturnsEntry = { value?: any }

export type MakeMockChainFnReturnType = {
  mockChainFn: Record<string, any>
  mocks: Record<string, jest.Mock>
  calls: CallsType[]
}

export type CallsType = CallFnType | CallValueType
export type CallFnType = { key: string | symbol; type: 'function'; args: any[] }
export type CallValueType = { key: string | symbol; type: 'value'; value?: any }
