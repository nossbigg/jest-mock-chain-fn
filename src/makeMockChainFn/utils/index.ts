export * from './makeNewMockMapping'
export * from './makeProxyHandler'
export * from './handlePropertyOverride'
export * from './clearExistingCallsHandler'
export * from './clearExistingMocksHandler'
