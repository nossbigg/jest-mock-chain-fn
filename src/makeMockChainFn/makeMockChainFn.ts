import { JestMockChainFnInternalState } from './constants'
import {
  BulilderMockInternalStateType,
  CallFnType,
  CallValueType,
  MakeBuilderMockOptions,
  MakeMockChainFnReturnType,
} from './typedefs'

export const makeMockChainFn = (
  options: MakeBuilderMockOptions = {}
): MakeMockChainFnReturnType => {
  const { mockPropertyReturns = {} } = options

  const builderMockInternalState: BulilderMockInternalStateType = {
    mocks: {},
    calls: [],
  }
  const { mocks, calls } = builderMockInternalState

  const builderMock = {
    [JestMockChainFnInternalState]: builderMockInternalState,
  }

  const makeNewMockMapping = (property: string, receiver: Object) => {
    const newMockFn = jest.fn()
    newMockFn.mockImplementation((...args) => {
      const callEntry: CallFnType = {
        key: property,
        type: 'function',
        args,
      }
      calls.push(callEntry)

      return receiver
    })

    mocks[property] = newMockFn
  }

  const builderMockHandler: ProxyHandler<any> = {
    get: function (_, _prop, receiver) {
      // allow access to internal state
      if (_prop === JestMockChainFnInternalState) {
        return builderMock[JestMockChainFnInternalState]
      }

      const property = _prop as string
      // allow property overrides
      if (property in mockPropertyReturns) {
        const mockPropertyReturnEntry = mockPropertyReturns[property]
        if ('value' in mockPropertyReturnEntry) {
          const { value } = mockPropertyReturnEntry

          const callEntry: CallValueType = {
            key: property,
            type: 'value',
            value,
          }
          calls.push(callEntry)

          return value
        }

        const callEntry: CallValueType = {
          key: property,
          type: 'value',
        }
        calls.push(callEntry)

        return receiver
      }

      // setup default method chaining mock, if missing
      if (!mocks[property]) {
        makeNewMockMapping(property, receiver)
      }

      // return fn mock
      return mocks[property]
    },
  }

  const mockChainFn = new Proxy(builderMock, builderMockHandler)
  return { mockChainFn, calls, mocks }
}
