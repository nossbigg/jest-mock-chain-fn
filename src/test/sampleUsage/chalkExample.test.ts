import { chalkExample } from './chalkExample'
import { makeMockChainFn } from '../../makeMockChainFn/makeMockChainFn'

describe('chalkExample', () => {
  const doTest = () => {
    const mockRedFn = jest.fn()
    mockRedFn.mockReturnValue('some-chalk-result')

    const mockPropertyReturns = {
      bold: {},
      red: { value: mockRedFn },
    }
    return makeMockChainFn({
      mockPropertyReturns,
    })
  }

  it('matches snapshot', () => {
    const { mockChainFn: chalk, calls } = doTest()

    chalkExample(chalk)
    expect(calls).toMatchSnapshot()
  })

  it('returns correct result', () => {
    const { mockChainFn: chalk } = doTest()

    const result = chalkExample(chalk)
    const expectedResult = 'some-chalk-result'
    expect(result).toEqual(expectedResult)
  })
})
