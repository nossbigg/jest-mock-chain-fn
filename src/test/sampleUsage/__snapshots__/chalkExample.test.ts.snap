// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chalkExample matches snapshot 1`] = `
Array [
  Object {
    "key": "bold",
    "type": "value",
  },
  Object {
    "args": Array [
      "some message",
    ],
    "key": "red",
    "type": "function",
  },
]
`;
